kernel: Change locking in k_thread_suspend()

peter-mitsis opened the pull request:
Updates k_thread_suspend() to work on SMP systems. It functions much like k_thread_abort().

If the target thread is running on another CPU, the target thread gets marked as "halting with a reason". Meanwhile the calling thread waits for the target thread to finish halting (suspend or abort). Note that the reason bit has been implemented such that an abort takes precedence over a suspend to cover the case a thread running on another CPU gets suspended and then aborted (or vice versa). If the target thread is not running on another CPU, then the target thread can be immediately aborted/suspended.

Fixes #58392

andyross commented:
Sorry, missed this earlier.  I guess I don't get the point here?  This is a behavior change: the call used to reschedule unconditionally if you suspended yourself (i.e. in the situation where we know it's required) and not if you suspended someone else.  Now it reschedules every time, which means that it acts as a yield point where it didn't before.

Personally I'm agnostic on that sort of thing, but there has been argument in the past about this sort of behavior.

But regardless: I don't see why this is a good idea?  Suspending another thread isn't an obvious spot where a user would expect a yield (because obviously we're the highest priority thread right now, so we can't be suspending anyone else running and thus the run choice shouldn't change on our local CPU).

My guess is that if there's a deadlock in the linked test that is fixed here, it's that the test is incorrectly sensitive to thread run order and needs better synchronization internally.  Effectively all this is doing is adding a `k_yield()` after the suspend, it's not fixing anything.

peter-mitsis commented:
On a UP system, the old code works great.  In such a system it makes perfect sense to only reschedule if the thread is attempting to suspend itself as the currently executing thread would by definition be the highest priority thread that is ready to run. On an SMP system, that assumption no longer holds. This means that on an SMP system, we must perform a reschedule as the thread that is being suspended could be executing on another core.

This in turn leads to the following choice--we can either have distinct logic for UP and SMP cases, or we can use the same logic for both. I opted for the same logic for both to keep things simple (and hopefully easier to maintain).

What are the penalties for unconditionally calling z_reschedule() in the UP case? Unless I am missing something, the only penalty should be a little extra time as it runs through the logic to determine whether or not to actually schedule in a new thread. Is that extra time problematic? I don't think so.

What about the scheduling point? Again, unless I am missing something, calling z_reschedule() should not result in any scheduling difference in UP.

andyross reviewed (changes requested):
OK, I think I'm seeing this.  This is deep, unfortunately.  Bear with me:

> What are the penalties for unconditionally calling z_reschedule() in the UP case?

If you're in a cooperative thread or holding a k_sched_lock(), it will break your lock and run another thread after this change, where it wouldn't before (or it would, but only in the case where you suspended yourself and would obviously expect/desire that behavior).  That's the source of previous arguments about scheduling points, mostly.  Our API is sensitive to this in ways that other OSes aren't, because we make promises they don't.  (The counterargument, FWIW, is that we've been historically bad at documenting where the scheduling points are, so they shouldn't count as a stable API).

> On an SMP system, that assumption no longer holds.

Sure it does.  The thread being suspended isn't running, so by definition suspending it isn't going to affect the run queue anywhere else either.

Except, as you point out, **what if it is actually running on another CPU**?

And here we're just falling down.  `k_thread_suspend()` simply isn't handling the case in SMP where you try to suspend an active thread.  It doesn't work, and it's never worked.  As I read it, all it would do is attempt to dequeue the thread (a noop since the run queue doesn't contain the actives threads in SMP) and then mark it suspended (again a noop since it doesn't do anything to stop the thread, albeit a dangerous one since the thread state bits are now inconsistent).

Doing this right requires signaling the other CPU with an IPI, then waiting for the thread to become inactive before returning.  That's more or less identical with the existing path for `k_thread_abort()` (which *does* support SMP) and could probably share the same infrastructure.

Alternatively, since this never worked, I think I'd personally be OK with just documenting the limitation[1] that trying to suspend an active thread on SMP platforms is illegal, and adding an assertion (or maybe even a panic?) to catch the cases that try.

But getting back to this PR and the bug: I really don't think this is a fix.  I think what's happened is that the modified run order is just fooling the test into a state where things come out "success".  I think this pushes me to a -1 on this particular change.

[1] As mentioned in other conversations, trying to asynchronously suspend another thread like this is inherently racy.   This IMHO isn't an API use case we should be going out of our way trying to support.

andyross commented:
> it will break your lock and run another thread

No, it won't. :)

I consistently make this mistake.  In fact z_reschedule() does not affect the preemption logic, you need to requeue a thread to make that happen, which is done in k_yield() but basically nowhere else.  So ignore that bit, this is vaguely undesirable from a consistency perspective (because apps are sometimes sensitive to run order when they shouldn't be) but is never a bug per se.

peter-mitsis commented:
You make a good, solid and understandable case for why the proposed fix was not actually a fix. Thank you! :)

peter-mitsis committed:
kernel: Add halt_queue field to k_thread

The halt queue will be used to identify threads that are waiting
for a thread on another CPU to finish suspending.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

andyross commented on kernel/thread.c:
```diff
@@ -273,12 +273,14 @@ static size_t copy_bytes(char *dest, size_t dest_size, const char *src, size_t s
+	static const char  *states_str[6] = {"dummy", "pending",
+					     "prestart", "dead",
+					     "suspended", "queued"};
+	static const size_t states_sz[6] = {5, 7, 8, 4, 9, 6};
```
Since you're in here reworking the logic anyway:  this is storing 9 bytes for each record (for the state value, pointer and the size), in addition to the string bytes themselves, which are never less than five bytes ("dead\0").  So just give it a 10 byte array (long enough for "suspended\0") indexed by the bit value (which is tight packed), and you actually save memory.  And the expression is simpler too as it doesn't require manually sorting or (yikes) calculating strlen() results.  Something like:  (note it does need an expression for the bit index of the states, right now all we have is masks in the headers)

````c
static const char states[][12] = {
    [DUMMY_BIT] = "dummy",
    [PENDING_BIT] = "pending",
    [PRESTART_BIT] = "prestart",
    [DEAD_BIT] = "dead",
    [SUSPENDED_BIT] = "suspended",
    [ABORTING_BIT] = "aborting",
    [QUEUED_BIT] = "queued",
};
````

The rest of the logic doesn't need to change beyond the obvious pointer/array swap.  Also the copy_bytes() can just turn into a strncpy or whatever since you don't need the size value any more.

andyross commented on kernel/sched.c:
```diff
@@ -657,24 +659,103 @@ void z_sched_start(struct k_thread *thread)
 }
 
+static void z_thread_halt(struct k_thread *thread, k_spinlock_key_t key,
+			  bool terminate)
```
Can you arrange this refactoring such that the renaming of z_thread_abort() to z_thread_halt() happens in place so it doesn't look like a big cut/paste patch?  Alternatively if that can't be arranged, can you split the patches to separate the motion of the code form the renaming?

peter-mitsis replied:
I'll try.

andyross reviewed (commented):
Pretty sure there's a bug with the state handling stemming from the "thread states can have more than one bit set now" mixup.  Also some nitpicking and suggestions.

But on the whole this looks like exactly how I hoped it would work, with a shared path for "stop this other thread on a different CPU so it can inspect scheduler state" that is only minimally different for the "async suspend" and abort cases.

(Though I remain on the record that any app trying to use k_thread_suspend() like this, **especially** ones on SMP platforms, remains inherently race-prone.  Robust unix apps have been warned off of using SIGSTOP for decades too).

cfriedt commented:
I'll give it a review once Andy's comments are addressed.

peter-mitsis commented on include/zephyr/kernel/thread.h:
```diff
@@ -343,6 +343,11 @@ struct k_thread {
 
+#ifdef CONFIG_SMP
+	/** threads waiting in k_thread_suspend() */
+	_wait_q_t  sync_queue;
```
Happy to do the rename.

On unifying "join_queue" and "halt_queue"  ...

Although the retry approach sounds appealing on the surface, it is the complexity surrounding corner cases make me hesitant to go there. If a suspend/resume wakes  threads waiting on k_thread_join(), presumably it is up to k_thread_join() to detect this and perform the retry if desired. This leaves a window where many things may happen. 

1. The timeout could have expired.
2. The target thread could have terminated (before or after the timeout).
3. The k_thread could have been re-used.
 
Getting the behaviour right is tricky.

Alternatively, we could have a smaller field in k_thread that indicates that the waiting thread is waiting to join or suspend and unpend accordingly. This would be cleaner than the retry, but I would expect it to be messier than the separate "halt_queue".

peter-mitsis commented on include/zephyr/kernel_structs.h:
```diff
@@ -59,8 +59,12 @@ extern "C" {
+
 /* Thread is being aborted */
-#define _THREAD_ABORTING (BIT(5))
+#define _THREAD_ABORTING (_THREAD_HALTING | BIT(6))
```
Perhaps I am too close to the code to see what you are seeing and my eyes are seeing what I want to see, but I am not seeing what you see.

The halting mask is two bits. The thread aborting is 2 bits (halt + reason). Both bits must be set for it to be an abort. k_thread_abort() tickles both bits. At the current commit, halting without aborting does not exist.

andyross replied:
Yeah, would be nicer to clean this up.  All the other _THREAD_xxx state values are single bits, and where we have conditions that are unions of multiple bit states we've always done that via a C API like `z_is_thread_prevented_from_running()`[1]

[1] (I don't like that name either, but the style is fairly clear: we have state bits and then we have predicates that interpret them, and they look different)

peter-mitsis commented on kernel/sched.c:
```diff
@@ -308,15 +308,21 @@ void z_requeue_current(struct k_thread *curr)
 static inline bool is_aborting(struct k_thread *thread)
 {
-	return (thread->base.thread_state & _THREAD_ABORTING) != 0U;
+	return (thread->base.thread_state & _THREAD_HALTING_MASK) == _THREAD_ABORTING;
```
At the current commit, halting without aborting does not exist.

peter-mitsis commented on kernel/sched.c:
```diff
@@ -1696,21 +1703,32 @@ static inline void unpend_all(_wait_q_t *wait_q)
+/**
+ * @brief Halt the specified thread
+ *
+ * Halt the specified thread and move it into the specified new state.
```
Happy to do so.

peter-mitsis commented:
I'm going to try reworking some of the commits. Based on the comments so far, not only is there is definitely room for improvement, but I think that the current breakdown has added some confusion.

peter-mitsis committed:
kernel: Rename end_thread() to halt_thread()

The routine halt_thread() acts nearly identical to end_thread()
except that instead of only halting the thread if the _THREAD_DEAD
state bit is not set, it will halt it if bit specified by the
parameter new_state is not set (which is always _THREAD_DEAD).

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Create z_thread_halt()

Extracts the essential thread synchronization logic when aborting
a thread from z_thread_abort() and moves it to its own routine
called z_thread_halt().

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Extend halting to support suspending

Extends the concept of halting a thread from just aborting a thread
to both aborting and suspending a thread.

Part of this involves updating k_thread_suspend() to operate in a
similar fashion to that of k_thread_abort().

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Update k_thread_state_str()

Updates k_thread_state_str() to interpret the halting bits
correctly.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis commented:
The two CI failures are not due to the proposed changes. I can replicate them locally using twister on the 'main' branch.

peter-mitsis commented:
(Resolved a merge conflict caused by a recent change to sched.c)

andyross commented on kernel/sched.c:
```diff
@@ -1752,48 +1759,37 @@ void z_thread_abort(struct k_thread *thread)
-
-	if (active) {
-		/* It's running somewhere else, flag and poke */
+	if (active_elsewhere) {
```
Simply to `if (thread_active_elsewhere(thread))` ?

andyross commented on kernel/sched.c:
```diff
@@ -1752,48 +1759,37 @@ void z_thread_abort(struct k_thread *thread)
 #ifdef CONFIG_SMP
-	if (is_aborting(thread) && thread == _current && arch_is_in_isr()) {
-		/* Another CPU is spinning for us, don't deadlock */
-		end_thread(thread);
```
Oh... actually this is wrong.  The comment is telling us why: we have a spin condition below where we're going to wait for another thread to die before we get to the end of the code and call end_thread().  But if that thread is itself in the same loop waiting on us to die, it's never going to reach end_thread() either and we'll deadlock.  So this can't be simplified, this has to be here.

I suspect the halt handling likewise has a similar deadlock that needs to be handled (i.e. two threads racing trying to suspend each other).

peter-mitsis replied:
I see what you mean. I think the logic needed here is ...

If the current thread is flagged as aborting, first call end_thread(_current) and then call z_swap() if at thread-level, or continue on if in an ISR.

andyross replied:
No need to handle the swap case, the IPI will preempt any thread-level logic anyway.  Honestly I think this code was fine as-is?

andyross reviewed (changes requested):
First patch introduces a deadlock.  Beyond that a few continuations of general review that probably shouldn't block merge.

peter-mitsis commented:
@andyross - Previously you wrote "Then I guess I'm not understanding what halting means? Why isn't abort a "subclass" of halt behavior, you do all the same stuff and then also terminate yourself?"

Aborting is indeed a sub-type of halting. Halting has two sub-types--aborting and suspending. When we don't care which we are dealing with, we can test with is_halting() which just tests the _THREAD_HALTING bit. When we do care, we test both bits.

peter-mitsis commented:
@andyross - It came to me last night as I was falling asleep what you were driving at, and what I was missing with the abort/suspend/halt bits. I think this minor tweak should make things more clear.

andyross commented on kernel/sched.c:
```diff
@@ -1752,48 +1759,49 @@ void z_thread_abort(struct k_thread *thread)
-	bool active = thread_active_elsewhere(thread);
+		if (!arch_is_in_isr()) {
+			z_swap(&sched_spinlock, key);
+			__ASSERT(false, "aborted _current back from dead");
```
So... this isn't a simplification anymore.  This is adding code to handle an extra case, which I guess is an optimization which catches the "aborted from another CPU" earlier, in the case where a thread is also self-aborting at the same time.    I guess I don't see the value there?   Why do we want to add code byte to make that particular operation faster?

And it seems needless to me anyway?   How would you reasonably hit this in practice?  The "aborting" state is flagged synchronously with the other CPU sending an IPI, this is only hit in the case where the current thread is managing to get into an (unlocked) abort call before the IPI arrives.  That seems almost impossible to hit in practice.

Really, can we just remove this first patch?  I honestly think it's hurting and not helping.  For sure, 100%, this is subtle code and takes time to understand.  But right now the changes mostly amount to your exploratory process of understanding and not any improvement on what's there.

peter-mitsis replied:
I am okay with dropping that commit.

nashif commented:
@andyross  can you please take another look?

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -961,6 +961,14 @@ int k_thread_cpu_pin(k_tid_t thread, int cpu);
+ * platforms, it is possible for application code to create a deadlock
+ * condition by simultaneously suspending or aborting using at least one
+ * suspension or termination from interrupt context. Zephyr cannot detect all
+ * such conditions.
```
> The old k_thread_abort() was deadlock-proof.

No, the old k_thread_abort() was NOT deadlock proof, and the current documentation explicitly states it. 

One such scenario wherein a deadlock can occur is if both CPU0 and CPU1 are in ISRs and they each try to abort the other CPU's interrupted thread. As both threads are currently active as they were interrupted, the _THREAD_ABORTING bit is set for both.  The next step in this sequence is for both of these CPUs to confirm that the thread they want to abort is marked as _THREAD_ABORTING and that it is not its current/interrupted thread. Since they are both in an ISR, they will each spin waiting for the thread to be aborted--deadlocked. Is this deadlock likely? No. But can it occur? Yes.

As the new proposed k_thread_suspend() shares the same infrastructure as k_thread_abort(), it too is vulnerable to deadlock. Thus, it must be mentioned.

I do have some ideas on how this unlikely deadlock can be prevented, but those are future plans.

andyross replied:
No, this is handled.  Or should be; "was designed to be", anyway.  I'm not going to hold this code up as a model of perfection or anything, but I will say that this (the second attempt to make abort SMP-safe) was excruciatingly hard to get right the first time, which informs my conservatism about this review.

There is a check inside the scheduler lock at the top of the abort path: the first abort-in-a-cycle thread to make it through the abort path with set the aborting flag and spin, the second will see the flag set and synchronously abort the current thread before spinning.  That guarantees that both interrupts exit k_thread_abort() in finite time, and the lock guarantees serialization of changes to the aborting flag such that if one CPU sets it the other is guaranteed not to have already check it.

It's possible there's a bug I'm not seeing or a glitch I'm forgetting.  But it's not that.

andyross replied:
I don't follow?  Specifically I don't see how you get to step 11 ("CPU1 ISR determines thread0 is not aborting"): that thread will have been affirmatively flagged aborting before the lock is released (step 10).

Really, can we just remove this comment?  I'm almost certain it's not correct that we have an unresolvable deadlock in k_sched_abort().  And even if it is correct, we should file that as a bug, write a reproducer and commit to fixing it instead of trying to document that kind of misfeature as a "intended undefined behavior" kind of thing.  Other OS's get this right, we should too.

andyross commented on kernel/sched.c:
```diff
@@ -665,24 +674,99 @@ void z_sched_start(struct k_thread *thread)
+			  bool terminate)
+{
+#ifdef CONFIG_SMP
+	if (is_halting(_current) && arch_is_in_isr()) {
```
Finally found some time to comb through.

This is a race in the current version: checking is_halting() here outside the lock doesn't prevent another CPU from setting it during the code below before the lock is taken, and indeed that can deadlock because now you're spinning on another thread while your current thread is flagged aborting, which is a forbidden state.  The old version was inside the lock.

peter-mitsis replied:
@andyross - Assuming the lock to which you are referring is spinlock **_sched_lock_**, then I am not seeing what you are seeing. The routine **_z_thread_halt()_** is only called from two places--both of which obtain the lock on **_sched_lock_** prior to the routine **_z_thread_halt()_** being called. We are therefore inside the lock when we perform the check on _**is_halting()**_. And as far as I can tell, the only places that modify either of the **_THREAD_HALTING** bits also have spinlock **_sched_lock_** locked.

If there are still concerns, perhaps we could schedule a time to discuss this (and of course summarise the results here).

peter-mitsis replied:
@andyross - ping

andyross commented on include/zephyr/kernel_structs.h:
```diff
@@ -60,9 +60,15 @@ extern "C" {
+#define _THREAD_SUSPENDING (BIT(6))
+
+/* Thread is in the process of halting */
+#define _THREAD_HALTING    (_THREAD_ABORTING | _THREAD_SUSPENDING)
```
Mentioned this long ago, but still don't love this API scheme "_THREAD_HALTING" looks like all the other enumerants, but it's not, it's a multi-bit mask and you can't use it like that.  Would prefer to see this "|" operator down inside "is_halting()" and e.g. "set_halting()", which is how other thread state predicates work.

peter-mitsis replied:
I missed doing that before. Will correct.

andyross commented on kernel/sched.c:
```diff
@@ -665,6 +666,69 @@ void z_sched_start(struct k_thread *thread)
+ * then it is safe to act immediately.
+ *
+ * @param thread Thread to suspend or abort
+ * @param key Current key for sched_spinlock
```
Can you more clearly document that this function will release the lock before returning?  That was the source of the last round of review mistakery.

peter-mitsis replied:
Will do.

andyross reviewed (commented):
Digging in on the deadlock doxygen bit (if you're absolutely certain this is a bug with current code let's file it and fix it, not document it).  Also restate a few nitpicks from earlier that would be nice to see fixed but don't necessesarily rise to a -1.

Beyond that you've worn me down and I agree this looks correct.  If you're willing to remove that one comment, commiters may feel free to consider my approval given implicitly.

peter-mitsis commented:
> Really, can we just remove this comment? I'm almost certain it's not correct that we have an unresolvable deadlock in k_sched_abort(). And even if it is correct, we should file that as a bug, write a reproducer and commit to fixing it instead of trying to document that kind of misfeature as a "intended undefined behavior" kind of thing. Other OS's get this right, we should too.

I will remove the comment from the doxygen from both k_thread_suspend() and k_thread_abort() as it exists in both. Issue #64646 has been filed to track this issue (assuming that it is real).

peter-mitsis commented:
This latest push should address the most recent comments.

andyross reviewed (approved):
\<Slaps a +1 on it.\>

\<Runs away.\>

cfriedt reviewed (approved):
It looks reasonable enough to me - do we have test coverage everywhere?
